File: test/src/test/java/cz/vhromada/common/test/facade/AbstractMovableChildFacadeTest.java
package cz.vhromada.common.test.facade;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

import java.util.List;

import cz.vhromada.common.Movable;
import cz.vhromada.common.converter.MovableConverter;
import cz.vhromada.common.facade.AbstractMovableChildFacade;
import cz.vhromada.common.facade.MovableChildFacade;
import cz.vhromada.common.service.MovableService;
import cz.vhromada.common.test.stub.AbstractMovableChildFacadeStub;
import cz.vhromada.common.test.stub.MovableStub;
import cz.vhromada.common.validator.MovableValidator;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

/**
 * A class represents test for class {@link AbstractMovableChildFacade}.
 *
 * @author Vladimir Hromada
 */
class AbstractMovableChildFacadeTest extends MovableChildFacadeTest<Movable, Movable, Movable, Movable> {

    /**
     * Instance of {@link Movable}
     */
    private Movable movable;

    /**
     * {@inheritDoc}
     * Cleanup movable.
     */
    @Override
    @BeforeEach
    public void setUp() {
        super.setUp();

        movable = null;
    }

    /**
     * Test method for {@link AbstractMovableChildFacade#AbstractMovableChildFacade(MovableService, MovableConverter, MovableValidator, MovableValidator)}
     * with null service for movable data.
     */
    @Test
    void constructor_NullMovableService() {
        assertThatThrownBy(() -> new AbstractMovableChildFacadeStub(null, getConverter(), getParentMovableValidator(), getChildMovableValidator()))
            .isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link AbstractMovableChildFacade#AbstractMovableChildFacade(MovableService, MovableConverter, MovableValidator, MovableValidator)}
     * with null converter for movable data.
     */
    @Test
    void constructor_NullConverter() {
        assertThatThrownBy(() -> new AbstractMovableChildFacadeStub(getService(), null, getParentMovableValidator(), getChildMovableValidator()))
            .isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link AbstractMovableChildFacade#AbstractMovableChildFacade(MovableService, MovableConverter, MovableValidator, MovableValidator)}
     * with null validator for parent data.
     */
    @Test
    void constructor_NullParentMovableValidator() {
        assertThatThrownBy(() -> new AbstractMovableChildFacadeStub(getService(), getConverter(), null, getChildMovableValidator()))
            .isInstanceOf(IllegalArgumentException.class);
    }

    /**
     * Test method for {@link AbstractMovableChildFacade#AbstractMovableChildFacade(MovableService, MovableConverter, MovableValidator, MovableValidator)}
     * with null validator for child data.
     */
    @Test
    void constructor_NullChildMovableValidator() {
        assertThatThrownBy(() -> new AbstractMovableChildFacadeStub(getService(), getConverter(), getParentMovableValidator(), null))
            .isInstanceOf(IllegalArgumentException.class);
    }

    @Override
    protected MovableChildFacade<Movable, Movable> getFacade() {
        return new AbstractMovableChildFacadeStub(getService(), getConverter(), getParentMovableValidator(), getChildMovableValidator());
    }

    @Override
    protected Movable newParentEntity(final Integer id) {
        return getMovable(id);
    }

    @Override
    protected Movable newParentDomain(final Integer id) {
        return getMovable(id);
    }

    @Override
    protected Movable newParentDomainWithChildren(final Integer id, final List<Movable> children) {
        return getMovable(id);
    }

    @Override
    protected Movable newChildEntity(final Integer id) {
        return getMovable(id);
    }

    @Override
    protected Movable newChildDomain(final Integer id) {
        return getMovable(id);
    }

    @Override
    protected Class<Movable> getParentEntityClass() {
        return Movable.class;
    }

    @Override
    protected Class<Movable> getParentDomainClass() {
        return Movable.class;
    }

    @Override
    protected Class<Movable> getChildEntityClass() {
        return Movable.class;
    }

    @Override
    protected Class<Movable> getChildDomainClass() {
        return Movable.class;
    }

    @Override
    protected void assertParentDeepEquals(final Movable expected, final Movable actual) {
    }

    /**
     * Returns movable object.
     *
     * @param id ID
     * @return movable object
     */
    private Movable getMovable(final Integer id) {
        if (movable == null) {
            movable = new MovableStub(id);
        } else {
            movable.setId(id);
        }

        return movable;
    }

}

File: Test/src/test/kotlin/cz/vhromada/common/test/facade/AbstractMovableChildFacadeTest.kt
package cz.vhromada.common.test.facade

import com.nhaarman.mockitokotlin2.KArgumentCaptor
import com.nhaarman.mockitokotlin2.any
import com.nhaarman.mockitokotlin2.argumentCaptor
import cz.vhromada.common.Movable
import cz.vhromada.common.facade.AbstractMovableChildFacade
import cz.vhromada.common.facade.MovableChildFacade
import cz.vhromada.common.test.stub.AbstractMovableChildFacadeStub
import cz.vhromada.common.test.stub.MovableStub
import org.junit.jupiter.api.BeforeEach

/**
 * A class represents test for class [AbstractMovableChildFacade].
 *
 * @author Vladimir Hromada
 */
class AbstractMovableChildFacadeTest : MovableChildFacadeTest<Movable, Movable, Movable, Movable>() {

    /**
     * Instance of [Movable]
     */
    private var movable: Movable? = null

    /**
     * {@inheritDoc}
     * Cleanup movable.
     */
    @BeforeEach
    override fun setUp() {
        super.setUp()

        movable = null
    }

    override fun getFacade(): MovableChildFacade<Movable, Movable> {
        return AbstractMovableChildFacadeStub(service, mapper, parentMovableValidator, childMovableValidator)
    }

    override fun newParentEntity(id: Int): Movable {
        return getMovable(id)
    }

    override fun newParentDomain(id: Int): Movable {
        return getMovable(id)
    }

    override fun newParentDomainWithChildren(id: Int, children: List<Movable>): Movable {
        return getMovable(id)
    }

    override fun newChildEntity(id: Int?): Movable {
        return getMovable(id)
    }

    override fun newChildDomain(id: Int?): Movable {
        return getMovable(id)
    }

    override fun getParentRemovedData(parent: Movable, child: Movable): Movable {
        return parent
    }

    override fun anyParentEntity(): Movable {
        return any()
    }

    override fun anyChildEntity(): Movable {
        return any()
    }

    override fun anyChildDomain(): Movable {
        return any()
    }

    override fun argumentCaptorParentDomain(): KArgumentCaptor<Movable> {
        return argumentCaptor()
    }

    override fun assertParentDeepEquals(expected: Movable, actual: Movable) {}

    /**
     * Returns movable object.
     *
     * @param id ID
     * @return movable object
     */
    private fun getMovable(id: Int?): Movable {
        if (movable == null) {
            movable = MovableStub(id)
        } else {
            movable!!.id = id
        }

        return movable!!
    }

}
